File: flutter_tex_js_android/android/src/main/kotlin/com/madlonkay/flutter_tex_js/TexRenderer.kt
package com.madlonkay.flutter_tex_js

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.os.Build
import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebView
import androidx.annotation.MainThread
import kotlinx.coroutines.*
import java.io.ByteArrayOutputStream
import kotlin.math.max
import kotlin.math.roundToInt

private const val html = """
<!DOCTYPE html>
<html id="root">
    <head>
        <!-- Viewport settings perhaps not needed on Android -->
        <meta name="viewport" content="initial-scale=1, maximum-scale=1, minimum-scale=1">
        <link rel="stylesheet" href="katex/katex.min.css">
        <script src="katex/katex.min.js"></script>
        <style type="text/css">
         body { background: transparent; margin: 0; }
         .katex-display { margin: 0; }
         .katex-html > .tag { position: unset !important; padding-left: 2em; }
         #math { float: left; }
        </style>
    </head>
    <body>
        <span id="math"></span>
    </body>
    <script>
     function getContainer() {
         return document.getElementById('math');
     }
     function render(math, displayMode) {
         try {
             katex.render(math, getContainer(), {
                 output: 'html',
                 displayMode: displayMode
             });
             // Returning immediately or setting 0 timeout here yielded bad results (incomplete
             // rendering, that weird zooming thing)
             const now = Date.now();
             setTimeout(function() { sendBounds(now); }, 100);
             return true;
         } catch (error) {
             sendError(error);
             return false;
         }
     }
     function setColor(color) {
         getContainer().style.color = color;
     }
     function setFontSize(fontSize) {
         getContainer().style.fontSize = fontSize;
     }
     function setNoWrap(noWrap) {
         getContainer().style.whiteSpace = noWrap ? 'nowrap' : 'unset';
     }
     function setWidth(width) {
         document.getElementById('root').style.width = width;
     }
     function sendError(error) {
         TexRenderer.onError(error.toString());
     }
     function sendBounds(timestamp) {
         const bounds = getContainer().getBoundingClientRect();
         TexRenderer.takeSnapshot(bounds.x, bounds.y, bounds.width, bounds.height, timestamp);
     }
     function loadAllFonts() {
         const fontLoadingPromises = [];
         for (const font of document.fonts) {
             fontLoadingPromises.push(font.load());
         }
         Promise.all(fontLoadingPromises).then(function() {
             TexRenderer.onReady();
         });
     }
     loadAllFonts();
    </script>
</html>
"""

@SuppressLint("SetJavaScriptEnabled", "AddJavascriptInterface")
class TexRenderer(private val context: Context) : CoroutineScope by MainScope() {

    private val density get() = context.resources.displayMetrics.density
    private val webView: WebView by lazy {
        WebView(context).apply {
            settings.javaScriptEnabled = true
            addJavascriptInterface(this@TexRenderer, "TexRenderer")
            // Size cannot be 0x0, but 1x1 works
            layout(0, 0, 1, 1)
            setBackgroundColor(Color.TRANSPARENT)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                setRendererPriorityPolicy(WebView.RENDERER_PRIORITY_IMPORTANT, false)
            }
        }
    }
    private var ready = false
    private var readyListener: (suspend () -> Unit)? = null
    private var resultListener: ((ByteArray?, TexRenderError?) -> Unit)? = null
    private var previousJs: String? = null
    private var previousBytes: ByteArray? = null

    init {
        if (BuildConfig.DEBUG && Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            WebView.setWebContentsDebuggingEnabled(true)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            WebView.enableSlowWholeDocumentDraw()
        }
    }

    @MainThread
    private suspend fun whenReady(completionHandler: suspend () -> Unit) = withContext(Dispatchers.Main) {
        if (ready) {
            completionHandler()
        } else {
            readyListener = completionHandler
            webView.loadDataWithBaseURL("file:///android_asset/", html, "text/html", null, null)
        }
    }

    @JavascriptInterface
    fun onReady() {
        Log.d("AMK", "onReady called; thread=${Thread.currentThread()}")
        ready = true
        val listener = readyListener!!
        launch {
            listener.invoke()
        }
        readyListener = null
    }

    @JavascriptInterface
    fun onError(error: String) {
        Log.d("AMK", "onError called; thread=${Thread.currentThread()}")
        val listener = resultListener!!
        launch {
            val err = TexRenderError("RenderError", "An error occurred during rendering", error)
            listener(null, err)
        }
        resultListener = null
        previousJs = null
        previousBytes = null
    }

    @JavascriptInterface
    fun takeSnapshot(x: Double, y: Double, width: Double, height: Double, timestamp: Double) {
        val xPx = (x * density).roundToInt()
        val yPx = (y * density).roundToInt()
        val widthPx = max((width * density).roundToInt(), 1)
        val heightPx = max((height * density).roundToInt(), 1)
        Log.d("AMK", "Taking snapshot of [$x, $y, $width, $height], scaled to [$xPx, $yPx, $widthPx, $heightPx]")
        val listener = resultListener!!
        launch {
            var byteArray: ByteArray? = null
            for (i in 0..100) {
                byteArray = getSnapshotBytes(xPx, yPx, widthPx, heightPx)
                if (byteArray.contentEquals(previousBytes)) {
                    // Log.d("AMK", "bytes were the same! delaying")
                    delay(50L)
                } else {
                    Log.d("AMK", "bytes look correct; took ${System.currentTimeMillis() - timestamp}ms over $i retries")
                    break
                }
            }
            previousBytes = byteArray
            listener.invoke(byteArray, null)
        }
        resultListener = null
    }

    private fun getSnapshotBytes(xPx: Int, yPx: Int, widthPx: Int, heightPx: Int): ByteArray {
        // val start = System.currentTimeMillis()
        val bitmap = Bitmap.createBitmap(xPx + widthPx, yPx + heightPx, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        webView.draw(canvas)
        val clipped = if (xPx != 0 || yPx != 0) {
            Bitmap.createBitmap(bitmap, xPx, yPx, widthPx, heightPx)
        } else {
            bitmap
        }
        val byteStream = ByteArrayOutputStream()
        clipped.compress(Bitmap.CompressFormat.PNG, 100, byteStream)
        val byteArray = byteStream.toByteArray()
        bitmap.recycle()
        clipped.recycle()
        // Log.d("AMK", "generating bytes took ${System.currentTimeMillis() - start}ms")
        return byteArray
    }

    @MainThread
    suspend fun render(math: String, displayMode: Boolean, color: String, fontSize: Double, maxWidth: Double, completionHandler: (ByteArray?, TexRenderError?) -> Unit) = withContext(Dispatchers.Main) {
        whenReady {
            if (resultListener != null) {
                val err = TexRenderError("ConcurrencyError", "A render job was already in progress", null)
                completionHandler(null, err)
                return@whenReady
            }
            val noWrap = maxWidth.isInfinite()
            val newWidth = if (noWrap) {
                "unset"
            } else {
                "${maxWidth}px"
            }
            val js = "setNoWrap($noWrap); setWidth('$newWidth'); setColor('$color'); setFontSize('${fontSize}px'); render('$math', $displayMode);"
            Log.d("AMK", "Executing JavaScript: $js")
            if (previousJs == js && previousBytes != null) {
                completionHandler.invoke(previousBytes!!, null)
            } else {
                resultListener = completionHandler
                previousJs = js
                // We call loadUrl instead of evaluateJavascript for backwards compatibility, and
                // because we can't really make use of the immediately returned result
                webView.loadUrl("javascript:$js")
            }
        }
    }
}

data class TexRenderError(val code: String, val message: String?, val details: String?)

File: flutter_tex_js_ios/ios/Classes/TexRenderer.swift
//
//  TexGenerator.swift
//  flutter_tex_js
//
//  Created by Aaron Madlon-Kay on 2020/07/01.
//

import Foundation
import WebKit

fileprivate let assetsUrl = Bundle(for: TexRenderer.self).url(forResource: "flutter_tex_js_katex", withExtension: "bundle")!

fileprivate let html = """
<!DOCTYPE html>
<html id="root">
    <head>
        <meta name="viewport" content="initial-scale=1, maximum-scale=1, minimum-scale=1">
        <link rel="stylesheet" href="katex.min.css">
        <script src="katex.min.js"></script>
        <style type="text/css">
         body { background: transparent; margin: 0; }
         .katex-display { margin: 0; padding: 1px 0; }
         .katex-html > .tag { position: unset !important; padding-left: 2em; }
         #math { float: left; padding-top: 1px; padding-bottom: 1px; }
        </style>
    </head>
    <body>
        <span id="math"></span>
    </body>
    <script>
     function getContainer() {
         return document.getElementById('math');
     }
     function render(math, displayMode) {
         try {
             katex.render(math, getContainer(), {
                 output: 'html',
                 displayMode: displayMode
             });
             return getBounds();
         } catch (error) {
             return error.toString();
         }
     }
     function setColor(color) {
         getContainer().style.color = color;
     }
     function setFontSize(fontSize) {
         getContainer().style.fontSize = fontSize;
     }
     function setNoWrap(noWrap) {
         getContainer().style.whiteSpace = noWrap ? 'nowrap' : 'unset';
     }
     function setWidth(width) {
         document.getElementById('root').style.width = width;
     }
     function getBounds() {
         return getContainer().getBoundingClientRect().toJSON();
     }
     function loadAllFonts() {
         const fontLoadingPromises = [];
         for (const font of document.fonts) {
             fontLoadingPromises.push(font.load());
         }
         Promise.all(fontLoadingPromises).then(function() {
             window.webkit.messageHandlers.ready.postMessage('ready');
         });
     }
     loadAllFonts();
    </script>
</html>
"""

enum TexError : Error {
    case engineError(message: String)
    case executionError(message: String)
    case pngConversion
    case concurrentRequest
}

class TexRenderer : NSObject, WKScriptMessageHandler {

    lazy var webView = initWebView()
    var ready = false
    var readyListener : (() -> Void)?
    var busy = false

    private func initWebView() -> WKWebView {
        let controller = WKUserContentController()
        controller.add(self, name: "ready")
        controller.add(self, name: "debug")

        let config = WKWebViewConfiguration()
        config.userContentController = controller

        let webView = WKWebView(frame: .zero, configuration: config)
        webView.isOpaque = false
        return webView
    }

    private func whenReady(_ completionHandler: @escaping () -> Void) {
        if ready {
            completionHandler()
        } else {
            readyListener = completionHandler
            webView.loadHTMLString(html, baseURL: assetsUrl)
        }
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        switch message.name {
        case "ready":
            ready = true
            readyListener?()
            readyListener = nil
        default:
            debugPrint("Message from WebView to: \(message.name); body: \(message.body)")
        }
    }

    func render(_ math: String, displayMode: Bool, color: String, fontSize: Double, maxWidth: Double, completionHandler: @escaping (Data?, Error?) -> Void) {
        whenReady { [weak self] in
            guard let self = self else {
                return
            }
            guard !self.busy else {
                completionHandler(nil, TexError.concurrentRequest)
                return
            }
            self.busy = true
            let noWrap = maxWidth.isInfinite
            let newWidth: String
            if noWrap {
                newWidth = "unset"
            } else {
                newWidth = "\(maxWidth)px"
            }
            let js = "setNoWrap(\(noWrap)); setWidth('\(newWidth)'); setColor('\(color)'); setFontSize('\(fontSize)px'); render('\(math)', \(displayMode));"
            log("Executing JavaScript: \(js)")
            self.webView.evaluateJavaScript(js) { [weak self] result, error in
                guard let self = self else {
                    return
                }
                if let result = result as? [String:Any] {
                    // Success
                    self.takeSnapshot(result, completionHandler)
                } else if let result = result as? String {
                    // Engine error
                    completionHandler(nil, TexError.engineError(message: result))
                } else if let result = result {
                    // Other error
                    completionHandler(nil, TexError.executionError(message: "\(result)"))
                } else {
                    completionHandler(nil, error)
                }
                self.busy = false
            }
        }
    }

    private func takeSnapshot(_ bbox: [String:Any], _ completionHandler: @escaping (Data?, Error?) -> Void) {
        let x = (bbox["left"] as! NSNumber).intValue
        let y = (bbox["top"] as! NSNumber).intValue
        let w = max((bbox["width"] as! NSNumber).intValue, 1)
        let h = max((bbox["height"] as! NSNumber).intValue, 1)
        let rect = CGRect(x: x, y: y, width: w, height: h)
        log("Taking snapshot of \(rect)")

        let snapConfig = WKSnapshotConfiguration()
        snapConfig.rect = rect
        if #available(iOS 13.0, *) {
            snapConfig.afterScreenUpdates = true
        }
        webView.takeSnapshot(with: snapConfig) { image, error in
            guard let image = image else {
                completionHandler(nil, error)
                return
            }
            guard let data = image.pngData() else {
                completionHandler(nil, TexError.pngConversion)
                return
            }
            completionHandler(data, nil)
        }
    }
}
